personal_finances/user/user_auth_exceptions.py

```diff
@@ -0,0 +1,100 @@
+from typing import List
+
+
+class AuthorizationHeaderNotPresent(Exception):
+    def __str__(self) -> str:
+        return "AuthorizationHeaderNotPresent"
+
+
+class AuthorizationHeaderEmptyContent(Exception):
+    def __str__(self) -> str:
+        return "AuthorizationHeaderEmptyContent"
+
+
+class InvalidAuthorizationHeader(Exception):
+    def __str__(self) -> str:
+        return "InvalidAuthorizationHeader"
+
+
+class InvalidLambdaEventInput(Exception):
+    def __str__(self) -> str:
+        return "InvalidLambdaEventInput"
+
+
+class UserNotFound(Exception):
+    def __str__(self) -> str:
+        return "UserNotFoundException"
+
+
+class PasswordNotMatch(Exception):
+    def __str__(self) -> str:
+        return "PasswordNotMatch"
+
+
+class InvalidUserPassword(Exception):
+    def __str__(self) -> str:
+        return "InvalidUserPassword"
+
+
+class InvalidUserName(Exception):
+    def __str__(self) -> str:
+        return "InvalidUserName"
+
+
+class ServerSecretNotFound(Exception):
+    def __str__(self) -> str:
+        return "InvalidUserName"
+
+
+class InvalidDynamoResponse(Exception):
+    def __str__(self) -> str:
+        return "InvalidUserName"
+
+
+_EXPCEPTION_TO_HTTP_RESPONSE_DICT: List[dict] = [
+    {
+        "statusCode": 401,
+        "body": "User or password incorrect",
+        "exceptions": [
+            UserNotFound,
+            PasswordNotMatch,
+        ],
+    },
+    {
+        "statusCode": 400,
+        "body": "Invalid authentication input",
+        "exceptions": [
+            AuthorizationHeaderNotPresent,
+            AuthorizationHeaderEmptyContent,
+            InvalidAuthorizationHeader,
+        ],
+    },
+    {
+        "statusCode": 400,
+        "body": "Username or password does not match the minimal security requirements",
+        "exceptions": [
+            InvalidUserPassword,
+            InvalidUserName,
+        ],
+    },
+    {
+        "statusCode": 500,
+        "body": "Internal Server Error",
+        "exceptions": [
+            InvalidLambdaEventInput,
+            ServerSecretNotFound,
+            InvalidDynamoResponse,
+        ],
+    },
+]
+
+
+def get_server_response_by_exception(exception: Exception) -> dict:
+    for http_response in _EXPCEPTION_TO_HTTP_RESPONSE_DICT:
+        exception_list = http_response["exceptions"]
+        if type(exception) in exception_list:
+            return {
+                "statusCode": http_response["statusCode"],
+                "body": http_response["body"],
+            }
```

Comment:
I like the clarity of exception type -> statuscode+message.

But anyone updating this dict needs to have some careful consideration to avoid bogus behavior by duplicating the exception in two places. At the very least there should be a comment indicating how to update this dict + and tests ensuring the intended consistency of this dict.

But a better alternative would be to create an index on the exception type, and raise an exception in case the dictionary is not consistent

We can do something like:
```python
_EXCEPTION_TO_HTTP_RESPONSE: Dict[Tuple, Dict] = {
    (UserNotFound, PasswordNotMatch): {"statusCode": 401, "body": "..."},
    (
        AuthorizationHeaderNotPresent,
        AuthorizationHeaderEmptyContent,
        InvalidAuthorizationHeader,
    ): {
        "statusCode": 400,
        "body": "...",
    },
}

_UNKNOWN_EXCEPTION_RESPONSE: Dict = {
    "statusCode": 500,
    "body": "...",
}


def _get_exception_type_index(exception_to_http_response) -> Dict[Exception, Dict]:
    exception_type_index = {}
    for exception_types, body in exception_to_http_response.items():
        if any(
                exception_type in exception_type_index
                for exception_type in exception_types
        ):
            LOGGER.warn(
                f"overriding response for one ore more of the exception type {exception_types}"
            )
            # There should be tests ensuring this is never reached in a shipped code.
            raise InconsistentExceptionDictionary
        exception_type_index = {
            **exception_type_index,
            **{exception_type: body for exception_type in exception_types}
        }

def get_server_response_by_exception(exception: Exception) -> dict:
    exception_type_index = _get_exception_type_index(_EXCEPTION_TO_HTTP_RESPONSE)
    try:
        return exception_type_index[exception]
    except KeyError:
        return _UNKNOWN_EXCEPTION_RESPONSE

```